org.spoofax.terms/src/org/spoofax/terms/TermFactory.java

```diff
@@ -115,9 +132,189 @@ public IStrategoTerm annotateTerm(IStrategoTerm term, IStrategoList annotations)
+            return buildInt(((IStrategoInt)term).intValue(), term, annotations);
+        } else if (TermUtils.isTuple(term)) {
+            return buildTuple(term.getAllSubterms(), term, annotations);
+        } else if (TermUtils.isReal(term)) {
+            return buildReal(((IStrategoReal)term).realValue(), term, annotations);
+        } else if (term instanceof IStrategoPlaceholder) {
+            return buildPlaceholder(((IStrategoPlaceholder)term).getTemplate(), term, annotations);
+        }
+
+        throw new UnsupportedOperationException(
+                "Unable to annotate term of type " + term.getClass().getName() + " in " + getClass().getName());
+    }
+
+    @Override
+    public IStrategoList.Builder createListBuilder(int initialCapacity, @Nullable IStrategoTerm replacee, @Nullable IStrategoList annotations) {
+        return new IStrategoList.Builder() {
+            private final StrategoArrayListBuilder innerBuilder = new StrategoArrayListBuilder(initialCapacity);
+
+            @Override
+            public void add(IStrategoTerm term) {
+                innerBuilder.add(term);
+            }
+
+            @Override
+            public IStrategoList build() {
+                IStrategoList newTerm = innerBuilder.build(annotations);
+                replaceStrategoTerm(newTerm, replacee);
+                return newTerm;
+            }
+
+            @Override
+            public boolean isEmpty() {
+                return innerBuilder.isEmpty();
+            }
+        };
+    }
+
+    /**
+     * Replaces the replacee with the new term.
+     *
+     * This implementation mutates the new term in-place.
+     *
+     * @param newTerm the new term
+     * @param replacee the term being replaced; or {@code null}
+     */
+    private void replaceStrategoTerm(IStrategoTerm newTerm, @Nullable IStrategoTerm replacee) {
+        if (replacee == null || newTerm == replacee) return;
+
+        preserveOrigins(newTerm, replacee);
+    }
+
+    /**
+     * Copied the origin of the replaced term to the new term, unless
+     * the new term already has an origin of its own.
+     *
+     * This implementation mutates the new term in-place.
+     *
+     * @param newTerm the new term
+     * @param replacee the term being replaced
+     */
+    private void preserveOrigins(IStrategoTerm newTerm, IStrategoTerm replacee) {
+        @Nullable IStrategoTerm curOrigin = OriginAttachment.getOrigin(newTerm);
+        if (curOrigin != null) {
+            // The term already has an origin of its own.
+            return;
+        }
+        @Nullable IStrategoTerm newOrigin = OriginAttachment.getOrigin(replacee);
+        if (newOrigin == null) {
+            // The replacee doesn't have an origin,
+            // so we bail out here to avoid creating a null origin attachment.
+            return;
+        }
+        // This mutates the term in-place.
+        OriginAttachment.setOrigin(newTerm, newOrigin);
+    }
```

Comment:
Given the current type hierarchy of term factories, this isn't supposed to be here. Instead the normal `TermFactory` doesn't know about attachments, and the `OriginTermFactory` does